event_sourcery/event_store/backend.py

```diff
@@ -0,0 +1,174 @@
+import abc
+from collections import UserDict
+from collections.abc import Iterator
+from contextlib import AbstractContextManager
+from functools import wraps
+from typing import NoReturn, Protocol, TypeVar, cast
+
+from typing_extensions import Self
+
+from event_sourcery.event_store import subscription
+from event_sourcery.event_store.dispatcher import Dispatcher, Listeners
+from event_sourcery.event_store.event import (
+    Encryption,
+    EventRegistry,
+    NoEncryptionStrategy,
+    NoKeyStorageStrategy,
+    RawEvent,
+    RecordedRaw,
+    Serde,
+)
+from event_sourcery.event_store.event_store import EventStore
+from event_sourcery.event_store.interfaces import (
+    EncryptionKeyStorageStrategy,
+    EncryptionStrategy,
+    OutboxFiltererStrategy,
+    OutboxStorageStrategy,
+    StorageStrategy,
+    SubscriptionStrategy,
+)
+from event_sourcery.event_store.outbox import Outbox
+from event_sourcery.event_store.tenant_id import DEFAULT_TENANT, TenantId
+
+
+class NoProviderConfigured(Exception):
+    pass
+
+
+def no_filter(entry: RawEvent) -> bool:
+    return True
+
+
+class NoOutboxStorageStrategy(OutboxStorageStrategy):
+    def outbox_entries(
+        self, limit: int
+    ) -> Iterator[AbstractContextManager[RecordedRaw]]:
+        return iter([])
+
+
+T_co = TypeVar("T_co", covariant=True)
+T = TypeVar("T")
+
+
+class _Provider(Protocol[T_co]):
+    def __call__(self, container: "_Container") -> T_co: ...
+
+
+def singleton(provider: _Provider[T]) -> _Provider[T]:
+    result: T | None = None
+
+    @wraps(provider)
+    def _wrapper(container: "_Container") -> T:
+        nonlocal result
+        if result is not None:
+            return result
+        result = provider(container)
+        return result
+
+    return _wrapper
+
+
+def not_configured(error_message: str) -> _Provider[T]:
+    def _raise(container: "_Container") -> NoReturn:
+        raise NoProviderConfigured(error_message)  # pragma: no cover
+
+    return _raise
+
+
+class _Container(UserDict[type[T], _Provider[T] | T]):
+    def __getitem__(self, _type: type[T] | type) -> T:
```

Comment:
Why do you need `type[T]` | type` ?

Reply:
It's an leftover from previous versions.

Reply:
Removed